packages/entity/src/internal/EntityDataManager.ts

```diff
@@ -236,4 +312,63 @@ export default class EntityDataManager<
     // TODO(wschurman): check for races with load
     await Promise.all(pairs.map(([key, value]) => this.invalidateOneAsync(key, value)));
   }
+
+  /**
+   * Invalidate all in-memory caches for sets of key-value pairs for all transactions and parent transactions.
+   * @param pairs - key-value pairs to invalidate
+   */
+  public invalidateKeyValuePairsForTransaction(
+    queryContext: EntityTransactionalQueryContext,
+    pairs: readonly LoadPair<TFields, TIDField, any, any, any>[],
+  ): void {
+    // TODO(wschurman): check for races with load
+
+    // invalidate all query contexts in transaction tree
+    const outermostTransactionalQueryContext =
+      EntityDataManager.getOutermostTransactionalQueryContextIfInNestedTransaction(queryContext);
+    const allQueryContextsToInvalidate = [
+      outermostTransactionalQueryContext,
+      ...EntityDataManager.getAllChildTransactionalQueryContexts(
+        outermostTransactionalQueryContext,
+      ),
+    ];
+    for (const currentQueryContext of allQueryContextsToInvalidate) {
+      for (const [key, value] of pairs) {
+        this.invalidateOneForTransaction(currentQueryContext, key, value);
+      }
+    }
+  }
+
+  /**
+   * Traverse to root of transactional query context tree.
+   */
+  private static getOutermostTransactionalQueryContextIfInNestedTransaction(
+    queryContext: EntityTransactionalQueryContext,
+  ): EntityTransactionalQueryContext {
+    if (queryContext.isInNestedTransaction()) {
+      return EntityDataManager.getOutermostTransactionalQueryContextIfInNestedTransaction(
+        queryContext.parentQueryContext,
+      );
+    } else {
+      return queryContext;
+    }
+  }
+
+  /**
+   * Get a list of all child query contexts recursively for a given query context.
+   */
+  private static getAllChildTransactionalQueryContexts(
```

Comment:
I would replace the word Child with Descendant because it's recursive all the way down.